### PR TITLE
Send email on access requests and changes.

### PR DESCRIPTION
**Pull request by larskendall**
When a user signs up or requests access, the department admin(s) should receive this email:

Screen Shot 2019-03-18 at 7 48 49 AM

When an **Approve** button is clicked in the email, the approving admin will automatically be brought to the **My Department** page, where the action will be carried out automatically and they will see this message for a few seconds before it fades:

Screen Shot 2019-03-18 at 7 51 04 AM

The user who was approved should receive this email when they're approved through the email or manually through the app:

Screen Shot 2019-03-18 at 7 49 16 AM

If the user's access request is rejected, they should receive this email:

Screen Shot 2019-03-18 at 7 49 45 AM

If a user's access is revoked once they've already been approved previously, they should receive this email:

Screen Shot 2019-03-18 at 7 49 59 AM

The department admin should see this message in the **My Department** page when rejecting a user's access request, similarly to when they approve the request:

Screen Shot 2019-03-18 at 7 51 37 AM

### CODE REVIEW

File: server/api/user/user.controller.js [hunk 1]
```diff
@@ -94,6 +103,144 @@ function sendWelcomeEmail(user) {
...
+      template_content: [],
+      message: {
+        merge: false,
+        merge_language: 'handlebars',
+        global_merge_vars: [{
+          name: 'DEPARTMENT_NAME',
+          content: department.name,
+        }, {
+          name: 'DEPARTMENT_IMAGE_URL',
+          content: `https://s3.amazonaws.com/statengine-public-assets/logos/${department.firecares_id}.png`,
```
Comment (chopchop505) [inline server/api/user/user.controller.js]: Unfortunately, this isn't standard.  It should be stored in department.logo_link

File: server/api/user/user.controller.js [hunk 2]
```diff
@@ -24,6 +24,15 @@ function handleError(res, statusCode) {
   };
 }
 
+async function getDepartmentAdmins(departmentId) {
+  return await User.findAll({
+    where: {
+      fire_department__id: departmentId,
```
Comment (chopchop505) [inline server/api/user/user.controller.js]: Not sure if it'll ever happen, but can we make sure departmentId is not undefined.  If it is, throw an error

File: server/api/user/user.controller.js [hunk 3]
```diff
@@ -94,6 +103,144 @@ function sendWelcomeEmail(user) {
...
+      template_content: [],
+      message: {
+        merge: false,
+        merge_language: 'handlebars',
+        global_merge_vars: [{
+          name: 'DEPARTMENT_NAME',
+          content: department.name,
+        }, {
+          name: 'DEPARTMENT_IMAGE_URL',
+          content: `https://s3.amazonaws.com/statengine-public-assets/logos/${department.firecares_id}.png`,
```
Comment (chopchop505) [inline server/api/user/user.controller.js]: Same comment

File: server/api/user/user.controller.js [hunk 4]
```diff
@@ -94,6 +103,144 @@ function sendWelcomeEmail(user) {
...
```
Comment (chopchop505) [inline server/api/user/user.controller.js]: Same comment

File: server/api/user/user.controller.js [hunk 5]
```diff
@@ -116,30 +263,39 @@ function addToMailingList(user) {
...
-        if(!req.body.requested_fire_department_id && !req.body.fire_department__id) {
-          sendWelcomeEmail(user)
-            .then(() => {
-              res.status(204).send({user});
-            });
-        } else {
-          res.status(204).send({user});
+  try {
+    await user.save();
+    addToMailingList(user);
```
Comment (chopchop505) [inline server/api/user/user.controller.js]: Can you double check this?  I would think we would want to await

File: server/api/user/user.controller.js [hunk 6]
```diff
@@ -116,30 +263,39 @@ function addToMailingList(user) {
...
+
+    if(!req.body.requested_fire_department_id && !req.body.fire_department__id) {
+      sendWelcomeEmail(user);
```
Comment (chopchop505) [inline server/api/user/user.controller.js]: await?

File: server/api/user/user.controller.js [hunk 7]
```diff
@@ -116,30 +263,39 @@ function addToMailingList(user) {
...
+    // Send access request to department admin if a department was set.
+    if(req.body.requested_fire_department_id) {
+      const department = await FireDepartment.find({
+        where: {
+          _id: req.body.requested_fire_department_id,
         }
-      }))
-    .catch(validationError(res));
+      });
+      sendRequestDepartmentAccessEmail(user, department);
```
Comment (chopchop505) [inline server/api/user/user.controller.js]: await?

File: server/api/user/user.controller.js [hunk 8]
```diff
@@ -116,30 +263,39 @@ function addToMailingList(user) {
...
+    }
+
+    res.status(204).send({ user });
+  } catch (err) {
+    validationError(res);
```
Comment (chopchop505) [inline server/api/user/user.controller.js]: this should probably be handleError to return 500

File: server/api/user/user.controller.js [hunk 9]
```diff
@@ -172,41 +328,59 @@ export function edit(req, res) {
...
-    .catch(handleError(res));
+
+  try {
+    await user.save();
+    const department = await FireDepartment.find({
+      where: {
+        _id: departmentId,
+      },
+    });
+    sendAccessRevokedEmail(user, department, hadAccess);
```
Comment (chopchop505) [inline server/api/user/user.controller.js]: await?

File: server/api/user/user.controller.js [hunk 10]
```diff
@@ -217,11 +391,18 @@ export function approveAccess(req, res) {
...
-    })
-    .catch(handleError(res));
+  try {
+    await user.save();
+    const department = await FireDepartment.find({
+      where: {
+        _id: user.fire_department__id,
+      },
+    });
+    sendAccessApprovedEmail(user, department, req.query.readonly);
```
Comment (chopchop505) [inline server/api/user/user.controller.js]: await?

**Comment by larskendall** [comment]
@chopchop505 I think this should be good to go now. I also fixed the email templates in Mandrill so that image aspect ratios will be preserved instead of forcing everything into a square.

**Pull request merged**